R/unique-keys.R

```diff
@@ -0,0 +1,283 @@
+    mutate(kind = "explicit UK")
+
+  out$uk_col <- new_keys(out$uk_col)
+  out
+}
+
+dm_get_all_implicit_uks_def_impl <- function(def, table = NULL, all_explicit) {
+  dm_get_all_fks_def_impl(def) %>%
+    select(table = parent_table, uk_col = parent_key_cols) %>%
+    distinct() %>%
+    # rm those that are PKs or explicit UKs
+    anti_join(all_explicit, c("table", "uk_col"))
+}
+
+
+#' Remove a unique key
+#'
+#' @description
+#' `dm_rm_uk()` removes one or more unique keys from a table and leaves the [`dm`] object otherwise unaltered.
+#' An error is thrown if no unique key matches the selection criteria.
+#' If the selection criteria are ambiguous, a message with unambiguous replacement code is shown.
+#' Foreign keys are never removed.
+#'
+#' @inheritParams dm_rm_pk
+#'
+#' @family primary key functions
+#'
+#' @return An updated `dm` without the indicated unique key(s).
+#'
+#' @export
+dm_rm_uk <- function(dm, table = NULL, columns = NULL, ...) {
+  dm_rm_uk_(dm, {{ table }}, {{ columns }}, ...)
+}
+
+dm_rm_uk_ <- function(dm, table, columns, ...) {
+  check_dots_empty()
+  check_not_zoomed(dm)
+
+  table_name <- dm_tbl_name_null(dm, {{ table }})
+  columns <- enexpr(columns)
+
+  dm_rm_uk_impl(dm, table_name, columns)
+}
+
+dm_rm_uk_impl <- function(dm, table_name, columns) {
+  def <- dm_get_def(dm)
+
+  if (is.null(table_name)) {
+    i <- which(map_int(def$uks, vec_size) > 0)
+  } else {
+    i <- which(def$table == table_name)
+    if (nrow(def$uks[[i]]) == 0) {
+      i <- integer()
+    }
+  }
+  if (length(i) == 0 && dm_is_strict_keys(dm)) {
+    abort_uk_not_defined()
+  }
+
+  ii <- if (!quo_is_null(columns)) {
+    ii_col <- map2(def$data[i], def$uks[i], ~ tryCatch(
+      {
+        vars <- eval_select_indices(columns, colnames(.x))
+        map_lgl(.y$column, ~ identical(names(vars), .x))
+      },
+      error = function(e) {
+        FALSE
+      }
+    ))
+
+    # if `columns` is not NULL, it refers to only one UK, therefore we can choose
+    # the first element of the list created by `map2()`
+
+    # FIXME: error message should be more informative: which UKs are available
+    # for the given table? What was the user input for `columns`?
+    if (!any(ii_col[[1]])) {
+      abort_uk_not_defined()
+    }
+    ii_col[[1]]
+  } else {
+    # if no `column` is provided by user, use all columns for matching
+    TRUE
+  }
+
+  # Talk about it
+  if (is.null(table_name) || quo_is_null(columns)) {
+    n_uk_per_table <- map_int(i, ~ nrow(def$uks[[.x]]))
+    message("Removing unique keys: %>%")
+    message("  ", glue_collapse(
+      glue(
+        "dm_rm_uk({tick_if_needed(rep(def$table[i], n_uk_per_table))}, {flatten_chr(map(i, ~ deparse_keys(def$uks[[.x]]$column)))})"
+      ), " %>%\n  "
+    ))
+  }
+  # Execute
+  # in case `length(i) > 1`: all tables have all their UKs removed, respectively
+  def$uks[i] <- if (length(i) > 1) {
+    list_of(new_uk())
+  } else {
+    list_of(filter(def$uks[[i]], !ii))
+  }
```

Comment:
I don't understand this code, but we'll notice if it causes trouble.

Reply:
`ii` contains the UK columns that are to be removed. We take the list of the current UKs and filter out `ii`. Well, as you say, let's see